services/galley/src/Galley/API/Teams.hs

```diff
@@ -471,42 +477,64 @@ updateTeamMember zusr zcon tid targetMember = do
   Log.debug $
     Log.field "targets" (toByteString targetId)
       . Log.field "action" (Log.val "Teams.updateTeamMember")
+
   -- get the team and verify permissions
   team <- tdTeam <$> (Data.team tid >>= ifNothing teamNotFound)
   user <-
     Data.teamMember tid zusr
       >>= permissionCheck SetMemberPermissions
+
   -- user may not elevate permissions
   targetPermissions `ensureNotElevated` user
-  Data.teamMember tid targetId >>= \case
-    Nothing -> do
+  previousMember <- Data.teamMember tid targetId >>= \case
+    Nothing ->
       -- target user must be in same team
       throwM teamMemberNotFound
-    Just previousMember -> do
-      when (downgradesOwner previousMember targetPermissions)
-        $
-        -- owner can be downgraded IFF it can be deleted
-        unless (canDeleteMember user previousMember)
-        $ throwM accessDenied
+    Just previousMember -> pure previousMember
+  when
+    ( downgradesOwner previousMember targetPermissions
+        && not (canDowngradeOwner user previousMember)
+    )
+    $ throwM accessDenied
+
   -- update target in Cassandra
   Data.updateTeamMember tid targetId targetPermissions
-  -- When journaling is enabled, we are guaranteed that teamMembersForFanout returns all users
+
+  -- Manage billing member data
+  when (team ^. teamBinding == Binding) $ do
+    -- add new billing member, if applicable
+    let isNewBillingMember =
+          not (isBillingMember previousMember)
+            && isBillingMember targetMember
+    when isNewBillingMember $
+      Data.addBillingTeamMember tid targetId
```

Comment:
I think this logic would be better suited inside the `updateTeamMember` function

Reply:
I moved it there, but I also removed the checks. I found that Cassandra doesn't complain if we insert a primary key twice, it just updates rest of the columns. In this table we only have the two primary key columns. I am not sure if this has any significant performance penalty, I couldn't find any answers online. Let me know if it is a bad idea.

Reply:
> I moved it there, but I also removed the checks. I found that Cassandra doesn't complain if we insert a primary key twice, it just updates rest of the columns. In this table we only have the two primary key columns. I am not sure if this has any significant performance penalty, I couldn't find any answers online. Let me know if it is a bad idea.

(I don't have much context, just replying to the comment)

Writing to cassandra using the same primary key is, as such, fine - however, if those requests happen close together time-wise, you may end up in a situation where the (INSERT key, value1, null, null) wins the race over the (INSERT key, value1, value2, value3), i.e. there can be race conditions in theory. At least I think this can be the case. Maybe I'm wrong. Running tests locally (with a bunch of inserts to local cassandra) can sometimes show this behaviour.

Reply:
This makes a lot of sense, thanks for the input, I have added the checks to make sure we insert/delete only when required.